src/gpu-tips-n-tricks.md

```diff
@@ -1,61 +1,293 @@
-
-Robert:
-I would say some of my “tips and tricks” are:
- 
-Avoid making use of GetData() it can lead to various issues later down the road when running things on the GPU
-Instead make use of HostRead(Write) in areas you know will stay on the Host and Read(Write) for kernels you know could end up on the GPU in the future.
-MFEM_FORALLs are an easy way to make sure your code can run on GPUs. However, they are limited to only being able to make use of public class functions within them.
-When passing data into the forall it’s a good idea to use Read(Write) to obtain the underlying data pointers passed into the macro
-For mfem::Vector and mfem::Array type classes, you can do something like obj.HostRead() or obj.Read() to pull the object’s data down to the host or device. You don’t actually need to make use of the pointers returned by these functions.
-Make use of device debug mode early on it. It’s a very useful tool in helping pinpoint whether or not your memory is in the wrong location when used within a compute kernel.
-If using device debug, if your code fails you can run gdb and set a breakpoint at mfem::mfem_error . The code will break as soon as it reaches this point and then you can backtrace from here to see what went wrong and where.
-Later, you can move the breakpoint to earlier points in the code and see if an object has memory currently located on the host or device.
-Defining the MFEM_TRACK_CUDA_MEM macro while building can help you see when data is transferred, allocated, and etc.
-If you see a lot of data movement between host and device that’s a good indication that you’re running a lot of host or device kernels that are making use of that data.
-You want to avoid this at all costs. Pin point where this is occurring and see if you can’t refactor your code so the data stays mainly on the device.
-Avoid mallocing memory on the GPU within frequently called kernels. CUDA malloc calls are slow and can hinder performance.
-If you really need to do such operations think of making use of a memory pool (e.g. Umpire) that way the mallocs are much cheaper on the GPU.
-If you know you’re going to use your mfem::Vector like object on the GPU go ahead call UseDevice(true) right after constructing the object.
-Be aware UseDevice() just returns a boolean tells you whether the object should be on the device or not. It’s not the same as UseDevice(true).
- 
-I’m sure I’ve got more but these are at least a start.
-
-Arturo:
-My recommendation would be to develop in debug mode, and set a break point here https://github.com/mfem/mfem/blob/master/general/error.cpp#L155
- 
-Most of my bugs stem from not moving the data correctly.
- 
-Also, I think it would be nice to have an overview of the flags in the memory manager class to help understand what is going on.
\ No newline at end of file
+# MFEM device handling
```

Comment:
I would suggest opening the entire document with a short example, maybe something like this:
```c++
Vector u;
Vector v;
// ...
auto u_data = u.Read();
auto v_data = v.ReadWrite();
MFEM_FORALL({i, u.Size(),
{
   v_data[i] *= u_data[i];
});
```
Something like this IMO explains better the basics of the device support in MFEM better than the explanation; after a reader has seen this example, then the explanations of the memory manager and the forall macros will make more sense.

Reply:
Thank you, it's a good suggestion, I added comments to make it even more clear, tell me what you think.